tensorboard/webapp/plugins/npmi/views/violin_filters/violin_filter/violin_filter_component.ts

```diff
@@ -0,0 +1,413 @@
+  }
+  get drawWidth(): number {
+    return this.chartWidth - this.drawMargin.left - this.drawMargin.right;
+  }
+  // Drawing containers
+  private svg: any;
+  private mainContainer: any;
+  private drawContainer: any;
+  // // Containers for axis and dots
+  private dotsGroup: any;
+  private yAxisGroup: any;
+  private xAxisGroup: any;
+  private miscGroup: any;
+  // Scales and axis
+  private xScale: d3.ScalePoint<string> = d3.scalePoint();
+  private xAxis?: d3.Axis<string>;
+  private yScale: d3.ScaleLinear<number, number> = d3.scaleLinear();
+  private yAxis?: d3.Axis<number | {valueOf(): number}>;
+  private xScaleNum: d3.ScaleLinear<number, number> = d3.scaleLinear();
+  // Brush
+  private brush: any;
+  // Misc
+  private nanLine: any;
+  private nanText: any;
+  private zeroLine: any;
+
+  private maxBinSize = 0;
+  private readonly rgbColors = ['240, 120, 80', '46, 119, 182', '190, 64, 36'];
+  private readonly area = d3
+    .area<ViolinBin>()
+    .x0(
+      function(this: ViolinFilterComponent, d: ViolinBin) {
+        return this.xScaleNum(-d.length);
+      }.bind(this)
+    )
+    .x1(
+      function(this: ViolinFilterComponent, d: ViolinBin) {
+        return this.xScaleNum(d.length);
+      }.bind(this)
+    )
+    .y(
+      function(this: ViolinFilterComponent, d: ViolinBin) {
+        if (d.x0! === -Infinity) {
+          return this.chartHeight - this.drawMargin.top;
+        }
+        return this.yScale((d.x1! + d.x0!) / 2.0);
+      }.bind(this)
+    )
+    .curve(d3.curveCatmullRom);
+
+  ngAfterViewInit(): void {
+    this.svg = d3.select(this.chartContainer.nativeElement).select('svg');
+    this.svg;
+    this.mainContainer = this.svg
+      .append('g')
+      .attr('transform', `translate(${this.margin.left}, ${this.margin.top})`);
+    this.drawContainer = this.mainContainer
+      .append('g')
+      .attr(
+        'transform',
+        `translate(${this.drawMargin.left}, ${this.drawMargin.top})`
+      );
+    this.dotsGroup = this.drawContainer.append('g').attr('class', 'dotsGroup');
+    this.yAxisGroup = this.mainContainer
+      .append('g')
+      .attr('class', 'axis axis--y');
+    this.xAxisGroup = this.mainContainer
+      .append('g')
+      .attr('class', 'axis axis--x');
+    this.miscGroup = this.drawContainer.append('g');
+    this.initializeBrush();
+    this.drawMisc();
+    this.redraw();
+  }
+
+  ngOnChanges(changes: SimpleChanges) {
+    if (this.svg) {
+      this.redraw();
+    }
+  }
+
+  redraw() {
+    this.setMaxBinSize();
+    this.updateAxes();
+    this.draw();
+  }
+
+  // Initializing/Updating the visualization props
+  setMaxBinSize() {
+    Object.values(this.chartData.violinData).forEach((dataElement) => {
+      const lengths = dataElement.map((bin) => bin.length);
+      const longest: number = Math.max(...lengths);
+      this.maxBinSize = Math.max(longest, this.maxBinSize);
+    });
+  }
+
+  updateAxes() {
+    this.xScale = d3
```

Comment:
Just wondering, do you need to reset the `this.xScale` over and over (when updateAxes get called) or can you just do `this.xScale.range(...).domain(...)` only?

Reply:
Indeed, thanks for the hint.